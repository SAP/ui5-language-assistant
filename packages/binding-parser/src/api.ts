export { parseBinding } from "./parser/index";
export { BindingParserTypes } from "./types";
export * from "./constant";
export {
  isAfterAdjacentRange,
  isBefore,
  isBeforeAdjacentRange,
  positionContained,
  rangeContained,
  isBindingExpression,
  isMetadataPath,
  isModel,
  isBindingAllowed,
  extractBindingSyntax,
} from "./utils";

import {
  Value,
  PrimitiveValue,
  CollectionValue,
  StructureValue,
} from "./types/binding-parser";

export const isCollectionValue = (
  value: Value | undefined
): value is CollectionValue => {
  if (!value) {
    return false;
  }
  if ((value as CollectionValue).type === "collection-value") {
    return true;
  }

  return false;
};

export const isStructureValue = (
  value: Value | undefined
): value is StructureValue => {
  if (!value) {
    return false;
  }
  if ((value as StructureValue).type === "structure-value") {
    return true;
  }

  return false;
};

/**
 * A value is considered as primitive if it is not structure or collection value
 */
export const isPrimitiveValue = (
  value: Value | undefined
): value is PrimitiveValue => {
  if (!value) {
    return false;
  }

  if (isCollectionValue(value) || isStructureValue(value)) {
    return false;
  }

  return true;
};
