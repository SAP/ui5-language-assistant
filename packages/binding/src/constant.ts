export const BINDING_ISSUE_TYPE = "binding-issue";
export const PROPERTY_BINDING_INFO =
  "sap.ui.base.ManagedObject.PropertyBindingInfo";
export const AGGREGATION_BINDING_INFO =
  "sap.ui.base.ManagedObject.AggregationBindingInfo";
export const Binding_Mode = "sap.ui.model.BindingMode";
