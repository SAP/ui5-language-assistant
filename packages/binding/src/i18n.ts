import type { i18n, TFunction } from "i18next";
import { DEFAULT_I18N_NAMESPACE } from "@ui5-language-assistant/context";
import i18nEn from "./i18n/i18n.json";
import i18next from "i18next";

export function initI18n(i18nInstance: i18n): void {
  i18nInstance.addResourceBundle(
    i18nInstance.language,
    DEFAULT_I18N_NAMESPACE,
    i18nEn
  );
}

export const adaptTranslatedText = (text: string): string => {
  if (/&quot;/.test(text)) {
    text = text.replace(/&quot;/g, '"');
  }
  if (/&#39;/.test(text)) {
    text = text.replace(/&#39;/g, "'");
  }
  if (/&amp;quot;/.test(text)) {
    text = text.replace(/&amp;quot;/g, "&quot;");
  }
  return text;
};

export const t: TFunction = (key: string, ...args) => {
  const result = i18next.t(key, ...args);
  return adaptTranslatedText(result);
};
