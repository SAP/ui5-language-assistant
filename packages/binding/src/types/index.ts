export type { BindingIssue } from "./issues";
export type {
  PropertyBindingInfoElement,
  Dependents,
  PropertyType,
  BindContext,
} from "./binding-info";
export { BindingInfoName, TypeKind } from "./binding-info";

export { BINDING_ISSUE_TYPE } from "../../src/constant";

export type {
  CursorContext,
  KeyValueContext,
  BaseContext,
  EmptyContext,
  InitialContext,
  KeyContext,
  ValueContext,
} from "./cursor";
