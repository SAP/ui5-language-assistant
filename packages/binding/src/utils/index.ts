export {
  typesToValue,
  valueTypeMap,
  isParts,
  findRange,
  defaultRange,
  getPropertyTypeWithPossibleValue,
} from "./element";

export { getCursorContext } from "./cursor";

export { getLogger } from "./logger";
