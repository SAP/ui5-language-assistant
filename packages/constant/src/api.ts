export {
  DEFAULT_OPEN_UI5_VERSION,
  DEFAULT_UI5_FRAMEWORK,
  DEFAULT_UI5_VERSION,
  DEFAULT_UI5_VERSION_BASE,
  OPEN_FRAMEWORK,
  UI5_FRAMEWORK_CDN_BASE_URL,
} from "./constant";
