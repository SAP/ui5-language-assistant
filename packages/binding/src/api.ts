import type { BindingIssue } from "./types";
import { BINDING_ISSUE_TYPE } from "./types";

export { getCompletionItems } from "./services/completion";
export { bindingValidators } from "./services/diagnostics";
export { getHover } from "./services/hover";
export type { BindingIssue } from "./types";

export function isBindingIssue<T extends { issueType: string }>(
  issue: BindingIssue | T
): issue is BindingIssue {
  return issue.issueType === BINDING_ISSUE_TYPE;
}

export {
  AGGREGATION_BINDING_INFO,
  PROPERTY_BINDING_INFO,
  FILTER_OPERATOR,
} from "./constant";

export { getBindingElements } from "./definition/definition";
