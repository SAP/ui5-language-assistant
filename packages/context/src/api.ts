import { getCustomViewId, getManifestDetails } from "./manifest";
import { getServices } from "./services";
import { Context } from "./types";
import { getSemanticModel } from "./ui5-model";
import { getYamlDetails } from "./ui5-yaml";

export {
  initializeManifestData,
  getManifestDetails,
  getUI5Manifest,
  findManifestPath,
} from "./manifest";
export {
  initializeUI5YamlData,
  getYamlDetails,
  getUI5Yaml,
  findYamlPath,
} from "./ui5-yaml";
export { getCDNBaseUrl } from "./utils";
export { cache } from "./cache";
export {
  reactOnCdsFileChange,
  reactOnUI5YamlChange,
  reactOnManifestChange,
  reactOnXmlFileChange,
  reactOnPackageJson,
} from "./watcher";

/**
 * Get context for a file
 * @param documentPath path to a file e.g. absolute/path/webapp/ext/main/Main.view.xml
 * @param modelCachePath path to a cached UI5 model
 */
export async function getContext(
  documentPath: string,
  modelCachePath?: string
): Promise<Context> {
  const manifestDetails = await getManifestDetails(documentPath);
  const yamlDetails = await getYamlDetails(documentPath);
  const ui5Model = await getSemanticModel(
    modelCachePath,
    yamlDetails.framework,
    manifestDetails.minUI5Version
  );
  const services = await getServices(documentPath);
  const customViewId = await getCustomViewId(documentPath);
  return { manifestDetails, yamlDetails, ui5Model, services, customViewId };
}
