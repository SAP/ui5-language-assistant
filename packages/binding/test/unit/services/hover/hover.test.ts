import { join } from "path";
import {
  Config,
  ProjectName,
  ProjectType,
  TestFramework,
  CURSOR_ANCHOR,
} from "@ui5-language-assistant/test-framework";
import { HoverService, getHoverService } from "../../helper";
import { initI18n } from "../../../../src/api";

describe("hover/index", () => {
  let getHover: HoverService;
  let framework: TestFramework;
  let root: string, documentPath: string;
  const viewFilePathSegments = [
    "app",
    "manage_travels",
    "webapp",
    "ext",
    "main",
    "Main.view.xml",
  ];
  beforeAll(async function () {
    const config: Config = {
      projectInfo: {
        name: ProjectName.cap,
        type: ProjectType.CAP,
        npmInstall: true,
        deleteBeforeCopy: false,
      },
    };
    framework = new TestFramework(config);
    const i18n = await framework.initI18n();
    initI18n(i18n);

    root = framework.getProjectRoot();
    documentPath = join(root, ...viewFilePathSegments);
    const uri = framework.getFileUri(viewFilePathSegments);
    getHover = getHoverService(
      framework,
      viewFilePathSegments,
      documentPath,
      uri
    );
  });
  describe("hover result", () => {
    it("on key", async () => {
      const snippet = `
        <Text text="{pa${CURSOR_ANCHOR}rts: ['some-value']}"></Text>
    `;
      const result = await getHover(snippet);
      expect(result).toMatchSnapshot();
    });
    it("on key - any type", async () => {
      const snippet = `
        <List
            items="{
                  path: '',
                  filters: [{
                      val${CURSOR_ANCHOR}ue1: ''
                  }]
              }"
        />
    `;
      const result = await getHover(snippet);
      expect(result).toMatchSnapshot();
    });
    it("on key inside collection", async () => {
      const snippet = `
        <Text text="{parts: ['some-test-data', { pa${CURSOR_ANCHOR}th: 'some-value'}]}"></Text>
    `;
      const result = await getHover(snippet);
      expect(result).toMatchSnapshot();
    });
    it("on key - inside structure", async () => {
      const snippet = `
      <List items="{
          path: '',
          filters: {
              pa${CURSOR_ANCHOR}th: ''
          }
      }" />
    `;
      const result = await getHover(snippet);
      expect(result).toMatchSnapshot();
    });
    it("on key - inside structure [reference]", async () => {
      const snippet = `
      <List items="{
          path: '',
          filters: {
              path: '',
              operator: 'All', 
              condition:{
                  filters: [{pa${CURSOR_ANCHOR}th: ''}]
              }
          }
      }" />
    `;
      const result = await getHover(snippet);
      expect(result).toMatchSnapshot();
    });
    it("on key - inside collection [filters]", async () => {
      const snippet = `
      <List items="{
          path: '',
          filters: [{
              pa${CURSOR_ANCHOR}th: ''
          }]
      }" />
    `;
      const result = await getHover(snippet);
      expect(result).toMatchSnapshot();
    });
    it("on key - inside collection [ nested filters]", async () => {
      const snippet = `
        <List
            items="{
                  path: '',
                  filters: [{
                      filters: [{
                        pa${CURSOR_ANCHOR}th: ''
                      }]
                  }]
              }"
        />
    `;
      const result = await getHover(snippet);
      expect(result).toMatchSnapshot();
    });
    it("on key - inside collection [ nested filters 02]", async () => {
      const snippet = `
        <List
            items="{
                  path: '',
                  filters: [{
                      filter${CURSOR_ANCHOR}s: [{
                      }]
                  }]
              }"
        />
    `;
      const result = await getHover(snippet);
      expect(result).toMatchSnapshot();
    });
    ["null", `''`, "0", "false"].forEach((value) => {
      it(`on key if ui5object has false value: ${value}`, async () => {
        const snippet = `
          <Text text="{ui5object: ${value}, pa${CURSOR_ANCHOR}th:'some-value' }" id="test-id"></Text>`;
        const result = await getHover(snippet);
        expect(result).toMatchSnapshot(value);
      });
    });
  });
  describe("no hover result", () => {
    it("check on value", async () => {
      const snippet = `
        <Text text="{parts: ['some${CURSOR_ANCHOR}-value']}"></Text>
    `;
      const result = await getHover(snippet);
      expect(result).toBeUndefined();
    });
    it("check on value inside collection", async () => {
      const snippet = `
        <Text text="{parts: ['some-test-data', { path: 'some-${CURSOR_ANCHOR}value'}]}"></Text>
    `;
      const result = await getHover(snippet);
      expect(result).toBeUndefined();
    });
    it("check on wrong key", async () => {
      const snippet = `
        <Text text="{par-${CURSOR_ANCHOR}ty: ['some-test-data', { path: 'some-value'}]}"></Text>
    `;
      const result = await getHover(snippet);
      expect(result).toBeUndefined();
    });
    it("check when property binding does not exits", async () => {
      const snippet = `
        <Text text="{par${CURSOR_ANCHOR}ty: ['some-test-data', { path: 'some-value'}]}"></Text>
    `;
      const result = await getHover(snippet, true);
      expect(result).toBeUndefined();
    });
    it("check not ui5 property", async () => {
      const snippet = `
        <List items="{inv${CURSOR_ANCHOR}oice>/Invoices}"> </List>
    `;
      const result = await getHover(snippet);
      expect(result).toBeUndefined();
    });
    it("check empty string", async () => {
      const snippet = `
        <Text text="${CURSOR_ANCHOR}"></Text>
     `;
      const result = await getHover(snippet);
      expect(result).toBeUndefined();
    });
    it("check binding expression", async () => {
      const snippet = `
        <Text text="{=  false ? 'Error' : 'Suc${CURSOR_ANCHOR}cess'}"></Text>
      `;
      const result = await getHover(snippet);
      expect(result).toBeUndefined();
    });
    it("check hover outside binding range", async () => {
      const snippet = `
        <Text text="{part: 'some-value'}, ${CURSOR_ANCHOR} {value: 'test-value'}"></Text>
     `;
      const result = await getHover(snippet);
      expect(result).toBeUndefined();
    });
    it("check binding is not allowed", async () => {
      const snippet = `
        <Text text="{i1${CURSOR_ANCHOR}8n>demo}"></Text>
     `;
      const result = await getHover(snippet);
      expect(result).toBeUndefined();
    });
    it("check exception", async () => {
      const snippet = `
        <Text text="{pa${CURSOR_ANCHOR}rts: ['some-value']}"></Text>
     `;
      const result = await getHover(snippet, false, true);
      expect(result).toBeUndefined();
    });
    it("on key if ui5object has truthy value", async () => {
      const snippet = `
          <Text text="{ui5object: true, pa${CURSOR_ANCHOR}th:'some-value' }" id="test-id"></Text>`;
      const result = await getHover(snippet, false, true);
      expect(result).toBeUndefined();
    });
  });
});
