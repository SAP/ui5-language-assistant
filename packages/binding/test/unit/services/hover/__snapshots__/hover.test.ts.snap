// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hover/index hover result on key - any type 1`] = `
Object {
  "contents": Object {
    "kind": "markdown",
    "value": "\`(class) sap.ui.model.Filter\`

---

**Type:** any

**Description:** First value to use with the given filter operator

**Optional:** true

[More information](https://ui5.sap.com/1.108.26/#/api/sap.ui.model.Filter)",
  },
}
`;

exports[`hover/index hover result on key - inside collection [ nested filters 02] 1`] = `
Object {
  "contents": Object {
    "kind": "markdown",
    "value": "\`(class) sap.ui.model.Filter\`

---

**Type:** sap.ui.model.Filter

**Description:** An array of filters on which the logical conjunction is applied

**Optional:** true

[More information](https://ui5.sap.com/1.108.26/#/api/sap.ui.model.Filter)",
  },
}
`;

exports[`hover/index hover result on key - inside collection [ nested filters] 1`] = `
Object {
  "contents": Object {
    "kind": "markdown",
    "value": "\`(class) sap.ui.model.Filter\`

---

**Type:** String

**Description:** Binding path for this filter

**Optional:** true

[More information](https://ui5.sap.com/1.108.26/#/api/sap.ui.model.Filter)",
  },
}
`;

exports[`hover/index hover result on key - inside collection [filters] 1`] = `
Object {
  "contents": Object {
    "kind": "markdown",
    "value": "\`(class) sap.ui.model.Filter\`

---

**Type:** String

**Description:** Binding path for this filter

**Optional:** true

[More information](https://ui5.sap.com/1.108.26/#/api/sap.ui.model.Filter)",
  },
}
`;

exports[`hover/index hover result on key - inside structure [reference] 1`] = `
Object {
  "contents": Object {
    "kind": "markdown",
    "value": "\`(class) sap.ui.model.Filter\`

---

**Type:** String

**Description:** Binding path for this filter

**Optional:** true

[More information](https://ui5.sap.com/1.108.26/#/api/sap.ui.model.Filter)",
  },
}
`;

exports[`hover/index hover result on key - inside structure 1`] = `
Object {
  "contents": Object {
    "kind": "markdown",
    "value": "\`(class) sap.ui.model.Filter\`

---

**Type:** String

**Description:** Binding path for this filter

**Optional:** true

[More information](https://ui5.sap.com/1.108.26/#/api/sap.ui.model.Filter)",
  },
}
`;

exports[`hover/index hover result on key 1`] = `
Object {
  "contents": Object {
    "kind": "markdown",
    "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`

---

**Type:** Array<(String | sap.ui.base.ManagedObject.PropertyBindingInfo)>

**Description:** Array of binding info objects for the parts of a composite binding; the structure of each binding info is the same as described for the <code>oBindingInfo</code> as a whole.

If a part is not specified as a binding info object but as a simple string, a binding info object will be created with that string as <code>path</code>. The string may start with a model name prefix (see property <code>path</code>).

<b>Note</b>: recursive composite bindings are currently not supported. Therefore, a part must not contain a <code>parts</code> property.

**Optional:** true

[More information](https://ui5.sap.com/1.108.26/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
  },
}
`;

exports[`hover/index hover result on key if ui5object has false value: '': '' 1`] = `
Object {
  "contents": Object {
    "kind": "markdown",
    "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`

---

**Type:** String

**Description:** Path in the model to bind to, either an absolute path or relative to the binding context for the corresponding model; when the path contains a '&gt;' sign, the string preceding it will override the <code>model</code> property and the remainder after the '&gt;' will be used as binding path

**Optional:** true

[More information](https://ui5.sap.com/1.108.26/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
  },
}
`;

exports[`hover/index hover result on key if ui5object has false value: 0: 0 1`] = `
Object {
  "contents": Object {
    "kind": "markdown",
    "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`

---

**Type:** String

**Description:** Path in the model to bind to, either an absolute path or relative to the binding context for the corresponding model; when the path contains a '&gt;' sign, the string preceding it will override the <code>model</code> property and the remainder after the '&gt;' will be used as binding path

**Optional:** true

[More information](https://ui5.sap.com/1.108.26/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
  },
}
`;

exports[`hover/index hover result on key if ui5object has false value: false: false 1`] = `
Object {
  "contents": Object {
    "kind": "markdown",
    "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`

---

**Type:** String

**Description:** Path in the model to bind to, either an absolute path or relative to the binding context for the corresponding model; when the path contains a '&gt;' sign, the string preceding it will override the <code>model</code> property and the remainder after the '&gt;' will be used as binding path

**Optional:** true

[More information](https://ui5.sap.com/1.108.26/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
  },
}
`;

exports[`hover/index hover result on key if ui5object has false value: null: null 1`] = `
Object {
  "contents": Object {
    "kind": "markdown",
    "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`

---

**Type:** String

**Description:** Path in the model to bind to, either an absolute path or relative to the binding context for the corresponding model; when the path contains a '&gt;' sign, the string preceding it will override the <code>model</code> property and the remainder after the '&gt;' will be used as binding path

**Optional:** true

[More information](https://ui5.sap.com/1.108.26/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
  },
}
`;

exports[`hover/index hover result on key inside collection 1`] = `
Object {
  "contents": Object {
    "kind": "markdown",
    "value": "\`(typedef) sap.ui.base.ManagedObject.PropertyBindingInfo\`

---

**Type:** String

**Description:** Path in the model to bind to, either an absolute path or relative to the binding context for the corresponding model; when the path contains a '&gt;' sign, the string preceding it will override the <code>model</code> property and the remainder after the '&gt;' will be used as binding path

**Optional:** true

[More information](https://ui5.sap.com/1.108.26/#/api/sap.ui.base.ManagedObject.PropertyBindingInfo)",
  },
}
`;
